auth-service/src/main/kotlin/uk/gov/communities/delta/auth/controllers/external/DeltaForgotPasswordController.kt

```diff
@@ -0,0 +1,185 @@
+package uk.gov.communities.delta.auth.controllers.external
+
+import com.google.common.base.Strings
+import io.ktor.server.application.*
+import io.ktor.server.request.*
+import io.ktor.server.response.*
+import io.ktor.server.routing.*
+import io.ktor.server.thymeleaf.*
+import org.slf4j.LoggerFactory
+import uk.gov.communities.delta.auth.config.*
+import uk.gov.communities.delta.auth.deltaRouteWithEmail
+import uk.gov.communities.delta.auth.services.*
+import javax.naming.NameNotFoundException
+
+class DeltaForgotPasswordController(
+    private val deltaConfig: DeltaConfig,
+    private val emailConfig: EmailConfig,
+    private val authServiceConfig: AuthServiceConfig,
+    private val ssoConfig: AzureADSSOConfig,
+    private val resetPasswordTokenService: ResetPasswordTokenService,
+    private val registrationSetPasswordTokenService: RegistrationSetPasswordTokenService,
+    private val userLookupService: UserLookupService,
+    private val emailService: EmailService,
+) {
+    private val logger = LoggerFactory.getLogger(this.javaClass)
+
+    fun forgotPasswordFormRoutes(route: Route) {
+        route.post {
+            forgotPasswordPost(call)
+        }
+        route.get {
+            call.respondForgotPasswordPage()
+        }
+    }
+
+    fun forgotPasswordEmailSentRoute(route: Route) {
+        route.get {
+            val emailAddress = call.request.queryParameters["emailAddress"]!!
+            call.respondEmailSentPage(emailAddress)
+        }
+    }
+
+    private suspend fun forgotPasswordPost(call: ApplicationCall) {
+        val formParameters = call.receiveParameters()
+        val emailAddress = formParameters["emailAddress"].orEmpty()
+
+        val message = if (Strings.isNullOrEmpty(emailAddress)) "An email address is required."
+        else if (!LDAPConfig.VALID_EMAIL_REGEX.matches(emailAddress)) "Must be a valid email address"
+        else null
+        if (message != null) return call.respondForgotPasswordPage(message, emailAddress)
+        logger.atInfo().addKeyValue("emailAddress", emailAddress).log("Forgot password request")
+
+        val ssoClientMatchingEmailDomain = ssoConfig.ssoClients.firstOrNull {
+            it.required && emailAddress.lowercase().endsWith(it.emailDomain)
+        }
+        if (ssoClientMatchingEmailDomain != null)
+            return call.respondRedirect(
+                deltaRouteWithEmail(
+                    deltaConfig.deltaWebsiteUrl,
+                    ssoClientMatchingEmailDomain.internalId,
+                    emailAddress
+                )
+            )
```

Comment:
Would be good to log so we can see how often it happens in practice

```suggestion
        if (ssoClientMatchingEmailDomain != null) {
            logger.atInfo().addKeyValue("ssoClient", ssoClientMatchingEmailDomain.internalId).log("Forgot password email matches required SSO domain, redirecting")
            return call.respondRedirect(
                deltaRouteWithEmail(
                    deltaConfig.deltaWebsiteUrl,
                    ssoClientMatchingEmailDomain.internalId,
                    emailAddress
                )
            )
        }
```